src/handle.rs

```diff
@@ -419,3 +436,282 @@ impl HandleHost {
+    if let Some(data) = self.get_pointer() {
+      Self::new_raw(
+        // SAFETY: We're in the isolate's thread, because Weak<T> isn't Send or
+        // Sync.
+        unsafe { self.isolate_handle.get_isolate_ptr() },
+        data,
+        finalizer,
+      )
+    } else {
+      Weak {
+        data: None,
+        isolate_handle: self.isolate_handle.clone(),
+      }
+    }
+  }
+
+  fn get_pointer(&self) -> Option<NonNull<T>> {
+    self.data.as_ref().and_then(|data| data.pointer.get())
+  }
+
+  pub fn is_empty(&self) -> bool {
+    self.get_pointer().is_none()
+  }
+
+  pub fn open<'a>(&'a self, isolate: &mut Isolate) -> Option<&'a T> {
+    if let Some(data) = self.get_pointer() {
+      let handle_host: HandleHost = (&self.isolate_handle).into();
+      handle_host.assert_match_isolate(isolate);
+      Some(unsafe { &*data.as_ptr() })
+    } else {
+      None
+    }
+  }
+
+  pub fn as_global(&self, isolate: &mut Isolate) -> Option<Global<T>> {
+    if let Some(data) = self.get_pointer() {
+      let handle_host: HandleHost = (&self.isolate_handle).into();
+      handle_host.assert_match_isolate(isolate);
+      Some(unsafe { Global::new_raw(isolate, data) })
+    } else {
+      None
+    }
+  }
+
+  pub fn as_local<'s>(
+    &self,
+    scope: &mut HandleScope<'s, ()>,
+  ) -> Option<Local<'s, T>> {
+    if let Some(data) = self.get_pointer() {
+      let handle_host: HandleHost = (&self.isolate_handle).into();
+      handle_host.assert_match_isolate(scope);
+      let local = unsafe {
+        scope.cast_local(|sd| {
+          v8__Local__New(sd.get_isolate_ptr(), data.cast().as_ptr()) as *const T
+        })
+      };
+      Some(local.unwrap())
+    } else {
+      None
+    }
+  }
+
+  // Finalization callbacks.
+  // TODO: These functions depend on T, but always behind a pointer. Make sure
+  // they're not monomorphized.
+
+  extern "C" fn first_pass_callback(wci: *const WeakCallbackInfo) {
+    // SAFETY: If this callback is called, then the weak handle hasn't been
+    // reset, which means the `Weak` instance which owns the pinned box that the
+    // parameter points to hasn't been dropped.
+    let weak_data = unsafe {
+      let ptr = v8__WeakCallbackInfo__GetParameter(wci);
+      &*(ptr as *mut WeakData<T>)
+    };
+
+    if let Some(data) = weak_data.pointer.take() {
+      unsafe { v8__Global__Reset(data.cast().as_ptr()) };
+    } else {
+      unreachable!();
+    }
+
+    // Only set the second pass callback if there's a finalizer.
+    if let Some(finalizer) = weak_data.finalizer.take() {
+      weak_data.finalizer.set(Some(finalizer));
+      unsafe {
+        v8__WeakCallbackInfo__SetSecondPassCallback(
+          wci,
+          Self::second_pass_callback,
+        )
+      };
+    }
+  }
+
+  extern "C" fn second_pass_callback(wci: *const WeakCallbackInfo) {
+    // FIXME!!!!: Do we know for a fact that the parameter hasn't been dropped??
```

Comment:
I think it is safe? `WeakData<T>` is not dropped until `Weak<T>` itself drops.

Reply:
The first pass callback must be called the moment the reference is GC'd, since it's that callback's job to reset the weak handle, but the second pass callback has no such guarantees as far as I'm aware. [Looking in the V8 code](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/handles/global-handles.cc;l=1567) it seems like sometimes second passes run during GC, but other times they're scheduled on the platform task runner (which I believe would only get called when pumping the V8 message loop at the beginning of the next run in Deno's event loop).

Reply:
What mechanism exists that could cause the `WeakData` to be dropped earlier than expected?

Reply:
My concern is a scenario where the value is GC'd while `Weak` is still alive, but the second pass callback is scheduled for later, and `Weak` drops before that second pass callback runs. We could have flags in `WeakData` to check whether the second callback has run, and whether the corresponding `Weak` has dropped, and then leak the `Box<WeakData>` in the drop impl and deallocate it in the second pass callback if this is the case.

It seems like the various ways to trigger GC in tests (`gc()`, sending a low-memory notification, even running the GC in predictable mode) will cause second passes to run synchronously, so I'm not sure that this is testable.